File: Codes/2131-longest-palindrome-by-concatenating-two-letter-words.cpp
/*
 * @Author: LetMeFly
 * @Date: 2025-05-25 23:39:17
 * @LastEditors: LetMeFly.xyz
 * @LastEditTime: 2025-05-26 00:04:19
 */
#if defined(_WIN32) || defined(__APPLE__)
#include "_[1,2]toVector.h"
#endif

class Solution {
public:
    int longestPalindrome(vector<string>& words) {
        int cnt[26][26] = {0};
        for (string& word : words) {
            cnt[word[0] - 'a'][word[1] - 'a']++;
        }
        int ans = 0, middle = 0;
        for (int i = 0; i < 26; i++) {
            ans += cnt[i][i] / 2 * 2;
            middle |= cnt[i][i] % 2;
            for (int j = i + 1; j < 26; j++) {
                ans += min(cnt[i][j], cnt[j][i]) * 2;
            }
        }
        return (ans + middle) * 2;
    }
};

File: Codes/2131-longest-palindrome-by-concatenating-two-letter-words.py
'''
Author: LetMeFly
Date: 2025-05-25 23:39:17
LastEditors: LetMeFly.xyz
LastEditTime: 2025-05-25 23:47:11
'''
from typing import List

class Solution:
    def longestPalindrome(self, words: List[str]) -> int:
        cnt = [[0] * 26 for _ in range(26)]
        for word in words:
            cnt[ord(word[0]) - ord('a')][ord(word[1]) - ord('a')] += 1
        ans = middle = 0
        for i in range(26):
            ans += cnt[i][i] // 2 * 2
            middle |= cnt[i][i] % 2
            for j in range(i + 1, 26):
                ans += min(cnt[i][j], cnt[j][i]) * 2
        return (ans + middle) * 2